components/ulp/esp32ulp_mapgen.py

```diff
@@ -2,71 +2,95 @@
+             */
+            #pragma once
+            #ifdef __cplusplus
+            extern "C" {{
+            #endif
+            """  # noqa: E222
+        ))
 
+    expr = re.compile('^\\s*\\d+: ([a-f0-9]{8})\\s+(\\d+) OBJECT\\s+GLOBAL\\s+DEFAULT\\s+[^ ]+ (.*)$')
     for line in f_sym:
-        # NM "posix" format output has the following structure:
-        # symbol_name symbol_type addr_hex [size_hex]
-        parts = line.split()
-        name = parts[0]
-        addr = int(parts[2], 16) + base_addr
-        f_h.write('extern uint32_t ulp_{0};\n'.format(name))
-        f_ld.write('PROVIDE ( ulp_{0} = 0x{1:08x} );\n'.format(name, addr))
+        # readelf format output has the following structure:
+        #   index: addr_hex     size TYPE  SCOPE DEFAULT    junk symbol_name
+        # So match the line with a regular expression to parse it first
+        groups = expr.match(line)
+        if groups is None:  # Ignore non global or non object
+            continue
+        addr = int(groups.group(1), 16) + base_addr
+        size = int(groups.group(2))
+        name = groups.group(3)
+        if not cpp_mode:
+            name = prefix + name
+        f_h.write('extern uint32_t {0}{1};'.format(name, '[{0}]'.format(int(size / 4)) if size > 4 else ''))
+        # Then set the address to find the variable
+        f_h.write(' asm(".equ {0}, 0x{1:08X}");\n'.format(name_mangling(prefix + name) if cpp_mode else name, addr))
 
-    f_h.write(textwrap.dedent(
-        """
-        #ifdef __cplusplus
-        }
-        #endif
-        """
-    ))
+    if cpp_mode:
+        f_h.write('}\n')
+    else:
+        f_h.write(textwrap.dedent(
+            """
+            #ifdef __cplusplus
+            }
+            #endif
+            """
+        ))
 
 
 def main() -> None:
     description = ('This application generates .h and .ld files for symbols defined in input file. '
                    'The input symbols file can be generated using nm utility like this: '
-                   '<PREFIX>nm -g -f posix <elf_file> > <symbols_file>')
+                   '<PREFIX>readelf -sW <elf_file> > <symbols_file>')
 
     parser = argparse.ArgumentParser(description=description)
     parser.add_argument('-s', '--symfile', required=True, help='symbols file name', metavar='SYMFILE', type=argparse.FileType('r'))
     parser.add_argument('-o', '--outputfile', required=True, help='destination .h and .ld files name prefix', metavar='OUTFILE')
     parser.add_argument('--base-addr', required=True, help='base address of the ULP memory, to be added to each symbol')
+    parser.add_argument('-p', '--prefix', required=False, help='prefix for generated header file', default='ulp_')
 
     args = parser.parse_args()
 
-    with open(args.outputfile + '.h', 'w') as f_h, open(args.outputfile + '.ld', 'w') as f_ld:
-        gen_ld_h_from_sym(args.symfile, f_ld, f_h, int(args.base_addr, 0))
+    with open(args.outputfile + '.h', 'w') as f_h:
```

Comment:
Should the file name be `.hpp` for C++ case, making it clear that the file shall not be usable from C code?

Reply:
Tried to do that but it's a PITA since changing the extension here means also making the CMake code conditional to the prefix type (make all dependency tracking & artifacts as variable, ...). I don't think it matters anyway, the only way to get a C++ header is to specify a prefix with a namespace declaration, and in that case, if the `.h` header is included in a C source file, the compiler will just choke on the "namespace" part so it's pretty obvious.